executor/executor.go

```diff
@@ -41,104 +50,140 @@ func New(items, concurrency int, metrics Metrics) *Executor {
+	keys state.Keys
 
 	dependencies set.Set[int]
 	blocking     set.Set[int]
 
 	executed bool
 }
 
-func (e *Executor) createWorker() {
-	e.wg.Add(1)
+func (e *Executor) runWorker() {
+	defer e.wg.Done()
 
-	go func() {
-		defer e.wg.Done()
+	for {
+		select {
+		case t, ok := <-e.executable:
+			if !ok {
+				return
+			}
 
-		for {
-			select {
-			case t, ok := <-e.executable:
-				if !ok {
-					return
-				}
-				if err := t.f(); err != nil {
-					e.stopOnce.Do(func() {
-						e.err = err
-						close(e.stop)
-					})
-					return
-				}
+			if err := t.f(); err != nil {
+				e.stopOnce.Do(func() {
+					e.err = err
+					close(e.stop)
+				})
+				return
+			}
 
-				e.l.Lock()
-				for b := range t.blocking { // works fine on non-initialized map
-					bt := e.tasks[b]
-					bt.dependencies.Remove(t.id)
-					if bt.dependencies.Len() == 0 { // must be non-nil to be blocked
-						bt.dependencies = nil // free memory
-						e.executable <- bt
-					}
-				}
-				t.blocking = nil // free memory
-				t.executed = true
-				e.completed++
-				if e.done && e.completed == len(e.tasks) {
-					// We will close here if there are unexecuted tasks
-					// when we call [Wait].
-					close(e.executable)
+			e.l.Lock()
+			// Update concurrent reads to only contain not-yet-executed reads
+			for k := range t.keys {
+				key := e.edges[k]
+				key.reads.Remove(t.id)
+			}
+			for b := range t.blocking { // works fine on non-initialized map
+				bt := e.tasks[b]
+				bt.dependencies.Remove(t.id)
+				if bt.dependencies.Len() == 0 { // must be non-nil to be blocked
+					bt.dependencies = nil // free memory
+					e.executable <- bt
 				}
-				e.l.Unlock()
-			case <-e.stop:
-				return
 			}
+			t.blocking = nil // free memory
+			t.executed = true
+			e.completed++
+			if e.done && e.completed == len(e.tasks) {
+				// We will close here if there are unexecuted tasks
+				// when we call [Wait].
+				close(e.executable)
+			}
+			e.l.Unlock()
+		case <-e.stop:
+			return
 		}
-	}()
+	}
 }
 
 // Run executes [f] after all previously enqueued [f] with
 // overlapping [conflicts] are executed.
-// TODO: Handle read-only/write-only keys (currently the executor
-// treats everything still as ReadWrite, see issue below)
-// https://github.com/ava-labs/hypersdk/issues/709
 func (e *Executor) Run(conflicts state.Keys, f func() error) {
 	e.l.Lock()
 	defer e.l.Unlock()
 
 	// Add task to map
 	id := len(e.tasks)
 	t := &task{
-		id: id,
-		f:  f,
+		id:   id,
+		f:    f,
+		keys: conflicts,
 	}
 	e.tasks[id] = t
 
 	// Record dependencies
-	for k := range conflicts {
-		latest, ok := e.edges[k]
+	for k, v := range conflicts {
+		key, ok := e.edges[k]
 		if ok {
-			lt := e.tasks[latest]
+			// No Allocate/Write has been requested yet
+			if key.allocateWrite == notSet {
+				if v == state.Read {
+					key.reads.Add(id)
+				} else {
+					key.allocateWrite = id
+					for b := range key.reads {
+						bt := e.tasks[b]
+						recordDependencies(t, bt)
```

Comment:
There is an error here where the `bt` may be executed but you are now adding dependencies that won't be cleared.